src/toxcore/tcp/codec.rs

```diff
@@ -109,13 +111,10 @@ impl Decoder for Codec {
     fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
         // deserialize EncryptedPacket
         let (consumed, encrypted_packet) = match EncryptedPacket::from_bytes(buf) {
-            IResult::Incomplete(_) => {
-                return Ok(None)
-            },
-            IResult::Error(error) => {
-                return Err(DecodeError::DeserializeEncryptedError { error, buf: buf.to_vec() })
-            },
-            IResult::Done(i, encrypted_packet) => {
+            Err(Err::Incomplete(_)) => return Ok(None),
+            Err(Err::Error(_)) => return Ok(None),
```

Comment:
I am not quite sure we have to return `Ok(None)` if we got nom::Err::Error. If we return `Ok(None)` we tell tokio::Codec to leave the buffer untouched and try to fill it with a new portion of bytes. So in some scenario we will get nom::Err::Error and we will return `Ok(None)` so the buffer will grow and grow and we will got a DOS.

Reply:
sorry for the delay. @kpp: `Err::Error` and `Err::Failure` have the same effect everywhere, except for combinators like `alt`. If one of the child parsers returns `Err:Error`, `alt` tries the next one, but if it returned `Err:Failure`, the error is returned right away, and we do not try any other branch. `Failure` indicates that we were in the right branch, but something went wrong. The `cut` combinator can be used to wrap another parser and transform its `Err::Error` results to `Err::Failure`